python_modules/dagster/dagster/_daemon/controller.py

```diff
@@ -266,6 +269,22 @@ def check_daemon_heartbeats(self) -> None:
                 " They may be running more slowly than expected or hanging."
             )
 
+    def check_workspace_freshness(self, last_workspace_update_time: float) -> float:
+        nowish = pendulum.now("UTC").float_timestamp
+        try:
+            if (nowish - last_workspace_update_time) > RELOAD_WORKSPACE_INTERVAL:
+                if self._grpc_server_registry:
+                    self._grpc_server_registry.clear_all_grpc_endpoints()
+                self._workspace_process_context.refresh_workspace()
+                return pendulum.now("UTC").float_timestamp
+        except Exception:
+            if (nowish - last_workspace_update_time) > DEFAULT_WORKSPACE_FRESHNESS_TOLERANCE:
+                self._logger.exception("Daemon controller surpassed workspace freshness tolerance.")
+                raise
+            else:
+                self._logger.exception("Daemon controller failed to refresh workspace.")
```

Comment:
nit: maybe  some "still within tolerance, continuing" type piece to this message